File: 0921. Minimum Add to Make Parentheses Valid.java
// Medium
// Stack, Counting
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/minimum-add-to-make-parentheses-valid/

class Solution {
  // 这题的意思是找：缺少多少括号才能让整个字符串有效
  // 可以用计数
  public int minAddToMakeValid(String s) {
      int left = 0, right = 0;
      for (char c : s.toCharArray()) {
          if (c == '(') {
              left++;
          } else {
              if (left > 0) {
                  left--;
              } else {
                  right++;
              }
          }
      }
      return left + right;
  }
}

File: 0501-1000/0921. Minimum Add to Make Parentheses Valid.py
# Medium
# Greedy
# Time:O(n), Space:O(1)
# https://leetcode.cn/problems/minimum-add-to-make-parentheses-valid/

class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        # left -> unmatch '('
        # right -> unmatch ')'
        left = right = 0

        for c in s:
            if c == '(':
                left += 1
            else:
                if left > 0:
                    left -= 1
                else:
                    right += 1
        
        return left + right
    